pkg/cmd/cli/cmd/version.go

```diff
@@ -0,0 +1,183 @@
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	clientcmdapi "k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	kubeversion "k8s.io/kubernetes/pkg/version"
+
+	"github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/cmd/util/tokencmd"
+	"github.com/openshift/origin/pkg/version"
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	versionLong = `
+Display client and server versions.`
+)
+
+type VersionOptions struct {
+	BaseName         string
+	Config           clientcmdapi.Config
+	ClientConfig     *restclient.Config
+	OClient          *client.Client
+	KClient          *kclient.Client
+	Out              io.Writer
+	ClientCmdFactory *clientcmd.Factory
+
+	PrintEtcdVersion    bool
+	PrintClientFeatures bool
+}
+
+// NewCmdVersion creates a command for displaying the version of this binary
+func NewCmdVersion(fullName string, f *clientcmd.Factory, out io.Writer, options VersionOptions) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "version",
+		Short: "Display client and server versions",
+		Long:  versionLong,
+		Run: func(cmd *cobra.Command, args []string) {
+			options.BaseName = fullName
+
+			if err := options.Complete(f, args, out); err != nil {
+				kcmdutil.CheckErr(kcmdutil.UsageError(cmd, err.Error()))
+			}
+
+			if err := options.RunVersion(); err != nil {
+				kcmdutil.CheckErr(err)
+			}
+		},
+	}
+
+	return cmd
+}
+
+func (o *VersionOptions) Complete(f *clientcmd.Factory, args []string, out io.Writer) error {
+	if len(args) > 0 {
+		return fmt.Errorf("no arguments should be passed")
+	}
+
+	o.Out = out
+	o.ClientCmdFactory = f
+
+	return nil
+}
+
+// RunVersion attempts to display client and server versions for Kubernetes and OpenShift
+func (o VersionOptions) RunVersion() error {
+	if o.ClientCmdFactory != nil {
+		fmt.Fprintf(o.Out, "%s\n", "CLIENTS")
+	}
+
+	fmt.Fprintf(o.Out, "%s %v\n", o.BaseName, version.Get())
+	fmt.Fprintf(o.Out, "kubernetes %v\n", kubeversion.Get())
+	if o.PrintEtcdVersion {
+		fmt.Fprintf(o.Out, "etcd %v\n", etcdversion.Version)
+	}
+
+	if o.PrintClientFeatures {
+		features := []string{}
+		if tokencmd.BasicEnabled() {
+			features = append(features, "Basic-Auth")
+		}
+		if tokencmd.GSSAPIEnabled() {
+			features = append(features, "GSSAPI")
+			features = append(features, "Kerberos") // GSSAPI or SSPI
+			features = append(features, "SPNEGO")   // GSSAPI or SSPI
+		}
+		fmt.Printf("features: %s\n", strings.Join(features, " "))
+	}
+
+	if o.ClientCmdFactory == nil {
+		return nil
+	}
+
+	fmt.Fprintf(o.Out, "\n%s\n", "SERVERS")
+
+	// max amount of time we want to wait for server to respond
+	var maxRequestDuration int32 = 10
+
+	done := make(chan struct{})
+	var serverVersionErr error
+	oVersion := ""
+	oVersionHost := ""
+	kVersion := ""
+	kVersionHost := ""
+
+	// start goroutine to fetch openshift / kubernetes server version
+	go func(f *clientcmd.Factory) {
```

Comment:
looks like every error here is fatal.  If that's the case, `done := make(chan error)` may be more appropriate.  Then below you can use 

```go
select{
    case err, closed := <- done:
        if !closed && err != nil{
            return err
```